pkg/controller/validatingadmissionpolicystatus/crd.go

```diff
@@ -0,0 +1,249 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validatingadmissionpolicystatus
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsv1client "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/cel/openapi/resolver"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+type crdTracker struct {
+	// affectedPolicies is the (CRD GroupResources) -> (affected policy names) mapping
+	affectedPolicies map[schema.GroupResource]sets.Set[string]
+	// policyDependencies is the (policy name) -> (required CRD GroupResource) mapping.
+	// It is the reverse of affectedPolicies.
+	policyDependencies map[string]sets.Set[schema.GroupResource]
+
+	// mu protects affectedPolicies and policyDependencies
+	mu sync.RWMutex
+
+	parent *Controller
+}
+
+// convertCRDValidationToSchema converts the openAPIV3Schema to a kube-openapi schema.
+// In case of nil validation or missing OpenAPIV3Schema, this function returns ErrSchemaNotFound.
+// The CustomResourceValidation object must be valid. Anything from the informer or client should have already passed
+// validation.
+// This function uses a JSON marshal-unmarshal conversion instead of Structural because
+//   - OpenAPIV3Schema is already validated and thus always a valid OpenAPI schema;
+//   - Structural involves private api-extensions API types. We have import-boss to prevent anything outside
+//     apiextensions-apiserver package to use them.
+func convertCRDValidationToSchema(validation *apiextensionsv1.CustomResourceValidation) (*spec.Schema, error) {
```

Comment:
This method of conversion is unfortunate. Is this for every CRD watched or is this only called after we've filtered out ones we don't need?

@apelisse @Jefftree  do we have any conversion utilities tucked away that I might not be aware of for this?

Reply:
This is only called for policies that a policy requires. The CRD informer only cares about the name of the incoming CRD.

Reply:
This is used to convert CRD schema to spec.Schema for validation and I also believe for OpenAPI. Not sure if there are any drawbacks for this use case:

https://github.com/kubernetes/kubernetes/blob/4b6c340c1396194135221c88b86314437fc86c6d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go#L206

Reply:
Unfortunately this function imports `apiextensions`, the private API package. import-boss will yell at me for importing `apiextensions`.

Reply:
OK, sounds like we should track this as a pre-GA fix. I don't see it as blocking while this is an off-by-default beta feature.

Reply:
Yeah, ping @sttts, we must be able to do better with this.

Reply:
If every CRD change requeues all the policies that intercept that CRD's type, does that mean we're reparsing/converting this schema N times for N policies for a single CRD update, or are we storing the converted schema form?

Reply:
The CRD will be converted N times for N policies. The structural/unstructured conversion and the schema conversion should be a main theme of pre-GA performance fine tuning.

Reply:
I've added a controller-level LRU cache for parsed schemas. Now each CRD should only be converted once a while (unless a policy refers to an unreasonable amount of CRDs to bust the cache).

Reply:
With https://github.com/kubernetes/kubernetes/pull/118260 (which unfortunatelly didn't make code freeze) we could expose helpers around the schema as they don't use internal types anymore.